helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/WagedInstanceCapacity.java

```diff
@@ -0,0 +1,164 @@
+
+    // reset the value of all capacity to 0.
+    for (String key : partCapacity.keySet()) {
+      partCapacity.put(key, 0);
+    }
+    return partCapacity;
+  }
+
+  /**
+   * Process the pending messages based on the Current states
+   * @param currentState - Current state of the resources.
+   */
+  public void processPendingMessages(CurrentStateOutput currentState) {
+    Map<String, Map<Partition, Map<String, Message>>> pendingMsgs = currentState.getPendingMessages();
+    
+    for (String resource : pendingMsgs.keySet()) {
+      Map<Partition, Map<String, Message>> partitionMsgs = pendingMsgs.get(resource);
+
+      for (Partition partition : partitionMsgs.keySet()) {
+        String partitionName = partition.getPartitionName();
+
+        // Get Partition Weight
+        Map<String, Integer> partCapacity = getPartitionCapacity(resource, partitionName);
+
+        // TODO - check
+        Map<String, Message> msgs = partitionMsgs.get(partition);
+        // TODO - Check
+        for (String instance : msgs.keySet()) {
+           reduceAvailableInstanceCapacity(instance, partCapacity);
+        }
+      }
+    }
+  }
+
+  /**
+   * Get the partition capacity given Resource and Partition name.
+   */
+  private Map<String, Integer> getPartitionCapacity(String resource, String partition) {
+    ClusterConfig clusterConfig = _cache.getClusterConfig();
+    ResourceConfig resourceConfig = _cache.getResourceConfig(resource);
+
+
+    // Parse the entire capacityMap from ResourceConfig
+    Map<String, Map<String, Integer>> capacityMap;
+    try {
+      capacityMap = resourceConfig.getPartitionCapacityMap();
+    } catch (IOException ex) {
+      return createDefaultParticipantWeight();
+    }
+    return WagedValidationUtil.validateAndGetPartitionCapacity(partition, resourceConfig, capacityMap, clusterConfig);
+  }
+
+  /**
+   * Get the instance remaining capacity.
+   * Capacity and weight both are represented as Key-Value.
+   * Returns the capacity map of available head room for the instance.
+   * @param instanceName - instance name to query
+   * @return Map<String, Integer> - capacity pair for all defined attributes for the instance.
+   */
+  public Map<String, Integer> getInstanceAvailableCapacity(String instanceName) {
+    return _perInstance.get(instanceName);
+  }
+
+  public boolean isInstanceCapacityAvailable(String instance, Map<String, Integer> partitionCapacity) {
+    Map<String, Integer> instanceCapacity = _perInstance.get(instance);
+    for (String key : instanceCapacity.keySet()) {
+      if (partitionCapacity.containsKey(key)) {
+        Integer partCapacity = partitionCapacity.get(key);
+        if (partCapacity != 0 && instanceCapacity.get(key) < partCapacity) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  public boolean reduceAvailableInstanceCapacity(String instance, Map<String, Integer> partitionCapacity) {
```

Comment:
nit: tag `@Override` for this and above.

Reply:
thanks.. i had forgotten to add in last commit. re-did it.